src/durable-cache/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+    pub fn get_local(&self, key: &C::Key) -> Option<&C::Val> {
+        self.local.get(key)
+    }
+
+    /// Get and return the value associated with `key`, syncing with the durable store if
+    /// necessary. If `key` does not exist, then a value is computed via `val_fn` and durably
+    /// stored in the cache.
+    pub async fn get(&mut self, key: &C::Key, val_fn: impl FnOnce() -> C::Val) -> &C::Val {
+        // Fast-path if it's already in the local cache.
+        // N.B. This pattern of calling `contains_key` followed by `expect` is required to appease
+        // the borrow checker.
+        if self.local.contains_key(key) {
+            return self.local.get(key).expect("checked above");
+        }
+
+        // Reduce wasted work by ensuring we're caught up to at least the
+        // pubsub-updated shared_upper, and then trying again.
+        self.sync_to(self.write.shared_upper().into_option()).await;
+        if self.local.contains_key(key) {
+            return self.local.get(key).expect("checked above");
+        }
+
+        // Okay compute it and write it durably to the cache.
+        let val = val_fn();
+        let mut expected_upper = self.local_progress;
+        loop {
+            let update = ((C::encode_key(key), C::encode_val(&val)), expected_upper, 1);
+            let new_upper = expected_upper + 1;
+            let ret = self
+                .write
+                .compare_and_append(
+                    &[update],
+                    Antichain::from_elem(expected_upper),
+                    Antichain::from_elem(new_upper),
+                )
+                .await
+                .expect("usage should be valid");
+            match ret {
+                Ok(()) => {
+                    self.sync_to(Some(new_upper)).await;
+                    return self.get_local(key).expect("just inserted");
+                }
+                Err(err) => {
+                    expected_upper = self.sync_to(err.current.into_option()).await;
+                    if self.local.contains_key(key) {
+                        return self.local.get(key).expect("checked above");
+                    }
+                    continue;
+                }
+            }
+        }
+    }
+
+    /// Durably set `key` to `value`. A `value` of `None` deletes the entry from the cache.
+    pub async fn set(&mut self, key: &C::Key, value: Option<&C::Val>) {
+        self.set_many(&[(key, value)]).await
+    }
+
+    /// Durably set multiple key-value pairs in `entries`. Values of `None` deletes the
+    /// corresponding entries from the cache.
+    pub async fn set_many(&mut self, entries: &[(&C::Key, Option<&C::Val>)]) {
+        let mut expected_upper = self.local_progress;
+        loop {
+            let mut updates = Vec::new();
+
+            for (key, val) in entries {
+                if let Some(prev) = self.local.get(key) {
+                    updates.push(((key, prev), expected_upper, -1));
+                }
```

Comment:
This will not work if there are any duplicate keys in `entries`. Hmm, This might have to take a `BTreeMap` which is annoying. Or maybe we just track seen keys in this method.

Reply:
I like the signature you have, so I'd be inclined to track seen keys. We're about to write stuff durably, so the extra HashSet alloc seems fine. You could also sort entries and look for dups that way

Probably worth a unit test for whatever the behavior is for the bad thing (panic?)

Reply:
Done